datasource/schema/schema.go

```diff
@@ -122,6 +125,130 @@ func (s Schema) TypeAtTerraformPath(ctx context.Context, p *tftypes.AttributePat
 	return fwschema.SchemaTypeAtTerraformPath(ctx, s, p)
 }
 
+// Validate verifies that the schema is not using a reserved field name for a top-level attribute.
+func (s Schema) Validate() diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	// Raise error diagnostics when data source configuration uses reserved
+	// field names for root-level attributes.
+	reservedFieldNames := map[string]struct{}{
+		"connection":  {},
+		"count":       {},
+		"depends_on":  {},
+		"lifecycle":   {},
+		"provider":    {},
+		"provisioner": {},
+	}
+
+	attributes := s.GetAttributes()
+
+	for k, v := range attributes {
+		if _, ok := reservedFieldNames[k]; ok {
+			diags.AddAttributeError(
+				path.Root(k),
+				"Schema Using Reserved Field Name",
+				fmt.Sprintf("%q is a reserved field name", k),
+			)
+		}
+
+		d := validateAttributeFieldName(path.Root(k), k, v)
+
+		diags.Append(d...)
+	}
+
+	blocks := s.GetBlocks()
+
+	for k, v := range blocks {
+		if _, ok := reservedFieldNames[k]; ok {
+			diags.AddAttributeError(
+				path.Root(k),
+				"Schema Using Reserved Field Name",
+				fmt.Sprintf("%q is a reserved field name", k),
+			)
+		}
+
+		d := validateBlockFieldName(path.Root(k), k, v)
+
+		diags.Append(d...)
+	}
+
+	return diags
+}
+
+// validFieldNameRegex is used to verify that name used for attributes and blocks
+// comply with the defined regular expression.
+var validFieldNameRegex = regexp.MustCompile("^[a-z0-9_]+$")
+
+// validateAttributeFieldName verifies that the name used for an attribute complies with the regular
+// expression defined in validFieldNameRegex.
+func validateAttributeFieldName(path path.Path, name string, attr fwschema.Attribute) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if !validFieldNameRegex.MatchString(name) {
+		diags.AddAttributeError(
+			path,
+			"Invalid Schema Field Name",
+			fmt.Sprintf("Field name %q is invalid, the only allowed characters are a-z, 0-9 and _. This is always a problem with the provider and should be reported to the provider developer.", name),
+		)
+	}
+
+	if na, ok := attr.(fwschema.NestedAttribute); ok {
+		nestedObject := na.GetNestedObject()
+
+		if nestedObject == nil {
+			return diags
+		}
+
+		attributes := nestedObject.GetAttributes()
+
+		for k, v := range attributes {
+			d := validateAttributeFieldName(path.AtName(k), k, v)
```

Comment:
Since the nested attributes might be done using list, map, or set mode, we'll also need to account for adding those intermediate steps (AtListIndex, AtMapKey, AtSetValue) to the path. We're not looking at a real configuration though, so it's a little awkward to make it "correct", although Terraform likely won't show the source configuration matching the path information when it is returned by the GetProviderSchema RPC.

Let's create a followup issue, but we should likely try to fix up this path information and consider including it in the diagnostic details in case it is not in Terraform's output. Maybe we can use something like the zero-value (e.g. `AtListIndex(0)`, `AtMapKey("")`, `AtSetValue(/* ? */)`) for describing those intermediate path steps for static validation.

Reply:
https://github.com/hashicorp/terraform-plugin-framework/issues/574